src/boot/fwupd.c

```diff
@@ -0,0 +1,576 @@
+        dp_size = device_path_size_with_limit(hdr, space_left);
+        if (space_left != dp_size) {
+                log_error("Variable '%ls' is too small for EFI_DEVICE_PATH: %zu != %zu", name, space_left, dp_size);
+                return EFI_INVALID_PARAMETER;
+        }
+
+        *ret_update = xnew0(FwupUpdateTable, 1);
+        (*ret_update)->name = xstrdup16(name);
+        (*ret_update)->size = size;
+        (*ret_update)->info = info;
+        (*ret_update)->attrs = flags;
+        TAKE_PTR(infop);
+
+        log_debug("Parsed UEFI update variable '%ls': flags=0x%08x, hw_inst=%" PRIu64 ", status=%" PRIu32,
+                  name,
+                  info->capsule_flags,
+                  info->hw_inst,
+                  info->status);
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS parse_update_vars(FwupUpdateTable ***updates, size_t *n_updates) {
+        size_t name_length = 1024 * sizeof(char16_t);
+        _cleanup_free_ char16_t *name = NULL;
+        EFI_STATUS err;
+
+        assert(updates);
+        assert(n_updates);
+
+        log_debug("Searching for UEFI updates...");
+
+        /* A series of fwupd-* named variables will be prepared by userspace, but there is no fixed name,
+         * so we need to iterate over the full table of variables and find them. Unfortunately there are no
+         * APIs to search by GUID, so it's really the entire table that has to be walked. */
+
+        name = xnew0(char16_t, name_length);
+
+        for (;;) {
+                EFI_GUID guid;
+
+                err = RT->GetNextVariableName(&name_length, name, &guid);
+                if (err == EFI_NOT_FOUND)
+                        break;
+                if (err == EFI_BUFFER_TOO_SMALL) {
+                        name = xrealloc(name, name_length, name_length * 2);
+                        name_length *= 2;
+                        continue;
+                }
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to get next variable name: %m");
+
+                if (!efi_guid_equal(&guid, MAKE_GUID_PTR(FWUPDATE)))
+                        continue;
+                if (!startswith_no_case(name, u"fwupd-"))
+                        continue;
+
+                log_debug("Found UEFI update variable '%ls'", name);
+
+                _cleanup_(fwup_update_table_freep) FwupUpdateTable *update = NULL;
+                err = parse_info_var(name, &update);
+                if (err != EFI_SUCCESS) {
+                        efivar_unset(MAKE_GUID_PTR(FWUPDATE), name, 0);
+                        return err;
+                }
+                if (!FLAGS_SET(update->info->status, FWUPDATE_ATTEMPT_UPDATE))
+                        continue;
+
+                EFI_TIME now;
+                err = RT->GetTime(&now, NULL);
+                if (err == EFI_SUCCESS)
+                        update->info->time_attempted = now;
+                update->info->status = FWUPDATE_ATTEMPTED;
```

Comment:
seems a bit too early to update this here, and seems a bit weird, that we leave the timestamp as it if GetTime() fails. souldn't this timestamp actually be taken right before we apply the capsule?

Reply:
This is how fwupd-efi implements it, so it is marked as attempted regardless if the setup fails later. And it makes ense, otherwise if something is broken in the setup for any reason (some malformed metadata?) you would end up looping forever. Instead by marking it as attempted (which doesn't mean successful) immediately, then it is tried only once, and userspace can deal with it later if it didn't work.

And the time is not reliable apparently, or so does `process_random_seed()` say, so I took that comment at face valule

Reply:
well, it doesn#t really make sense to patch these buffers here, because this code does not update the variable. that happens much later. All I am saying is to moev this down, where we are actually write the efi var, i.e. below where we set the things via efivar_set_raw()...